README
Copyright 1991, 1996, 1999, 2000 Free Software Foundation, Inc.

Copyright 2008, 2009 William Hart

This file is part of the MPIR Library.

The MPIR Library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 3 of the License, or (at your
option) any later version.

The MPIR Library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with the MPIR Library; see the file COPYING.LIB.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301, USA.






			THE MPIR LIBRARY

MPIR is a fork of the GNU Multi Precision library (GMP -- see http://gmplib.org)

MPIR is a library for arbitrary precision arithmetic, operating on signed
integers, rational numbers, and floating point numbers.  It has a rich set of
functions, and the functions have a regular interface.

MPIR is designed to be as fast as possible, both for small operands and huge
operands.  The speed is achieved by using fullwords as the basic arithmetic
type, by using fast algorithms, with carefully optimized assembly code for the
most common inner loops for lots of CPUs, and by a general emphasis on speed
(instead of simplicity or elegance).

GMP/MPIR is believed to be faster than any other similar library.  Its advantage
increases with operand sizes for certain operations, since MPIR in many
cases has asymptotically faster algorithms.

MPIR is free software and may be freely copied on the terms contained in the
files COPYING.LIB and COPYING (most of MPIR is under the former, some under
the latter).



			OVERVIEW OF MPIR

There are four classes of functions in MPIR.

 1. Signed integer arithmetic functions (mpz).  These functions are intended
    to be easy to use, with their regular interface.  The associated type is
    `mpz_t'.

 2. Rational arithmetic functions (mpq).  For now, just a small set of
    functions necessary for basic rational arithmetics.  The associated type
    is `mpq_t'.

 3. Floating-point arithmetic functions (mpf).  If the C type `double'
    doesn't give enough precision for your application, declare your
    variables as `mpf_t' instead, set the precision to any number desired,
    and call the functions in the mpf class for the arithmetic operations.

 4. Positive-integer, hard-to-use, very low overhead functions are in the
    mpn class.  No memory management is performed.  The caller must ensure
    enough space is available for the results.  The set of functions is not
    regular, nor is the calling interface.  These functions accept input
    arguments in the form of pairs consisting of a pointer to the least
    significant word, and an integral size telling how many limbs (= words)
    the pointer points to.

    Almost all calculations, in the entire package, are made by calling these
    low-level functions.

For more information on how to use MPIR, please refer to the documentation.
It is composed from the file mpir.texi, and can be displayed on the screen or
printed.  How to do that, as well how to build the library, is described in
the INSTALL file in this directory.



			REPORTING BUGS

If you find a bug in the library, please make sure to tell us about it!

You should first check the MPIR web pages at http://www.mpir.org/. There will 
be patches for all known serious bugs there.

Report bugs to our development list: http://groups.google.com/group/mpir-devel.  
What information is needed in a good bug report is described in the manual.  
The same address can be used for suggesting modifications and enhancements.




----------------
Local variables:
mode: text
fill-column: 78
End:


NEWS
Copyright 1996, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006 Free Software
Foundation, Inc.

Copyright 2009, 2010 William Hart

This file is part of the MPIR Library.

The MPIR Library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at your
option) any later version.

The MPIR Library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with the GNU MP Library; see the file COPYING.LIB.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
MA 02110-1301, USA.

Changes between MPIR 3.0.0 and MPIR 2.7.2

Bug Fixes:

* Port bug fix for gmp_get_str from GMP
* Fix t-locale failure on Windows
* Attempt to fix issue with gmp_randinit_set for _mt generator
* Fix long/long long pointer issue on Windows 64
* Clz_tab array length to 129 in line with GMP

Features:

* New Intel Skylake assembly support due to Jens Nurmann, Alex Kruppa and GMP
* New Intel Haswell assembly support due to Alex Kruppa and GMP
* Rudimentary Broadwell support (not optimisation)
* Improved AMD Bulldozer support due to Alex Kruppa
* Faster mpz_powm, mpz_powm_ui from GMP
* New mpz_limbs functionality from GMP 6
* New mpn_sizeinbase, mullow_basecase_n, binvert, redc_1, redc_2, redc_n functions from GMP 6
* New mpn_nsumdiff_n function (speeds up FFT on haswell)
* Visual Studio 2017 support
* mpir.net for interface to .net languages
* Appveyor-CI support

Other:

* Separate yasm from MPIR build (use --with-yasm=/path/to/yasm with configure)
* Remove autogenerated Makefiles, makeinfo and configure from version control

Changes between MPIR 2.7.2 and MPIR 2.7.1

Bug Fixes:

* Increase .so version numbering to prevent conflict with MPIR 2.7.1

Features:

* Wave motion gun

Changes between MPIR 2.7.1 and MPIR 2.7.0

Bug Fixes:

* Fix bug in Karatsuba assembly code on Windows
* Fix bug in basecase division code (by removing implementation)

Features:

* Add building of tests to command line build for Windows
* Improve speed and tune targets on Windows
* Various MSVC build improvements

Other:

* Add some missing information in AUTHORS
* Travis continuous integration

Changes between MPIR 2.7.0 and MPIR 2.6.0

Bug Fixes:

* Many build fixes
* Fix fat binary build on 32 bit x86
* Fix minor bugs in invert.c and t-invertc.
* Remove broken assert in generic build
* Fix bug in FFT that causes tuning code to crash
* fix bug in sqr_basecase.asm on Windows
* Fix bug in mpz_nextprime
* Fix aliasing bug in mpz_nthroot
* Fix bug in mpn_mulmod_basecase
* Add missing GMP_DECLSPEC's causing segfaults on Windows

Speedups:

* faster divide-and-conquer division using new algorithm of W. Hart
* merge many speedups for number theoretic functions from GMP project
  (factorial, multi-factorial, primorial, n-choose-k, jacobi symbols)
* fft versions of mpn_mulmod_bnm1/p1 to speed up GMP-ECM
* switch to using the much cleaner gcdext implementation from GMP

Features:

* Cygwin64 support
* MSYS2 support
* configure support for latest Intel/AMD chips
* MSVC 2015 support
* Clean up Jason Moxham's old C code
* change handling of stdint.h (user optionally includes it before mpir.h)
* support OpenBSD 5.6 and later
* GCC 4.9 and 5.x support
* Clang support
* C++11 C++ compiler support

Changes between MPIR 2.6.0 and MPIR 2.5.2

Bug Fixes:

* fixed bug in speed/common for dc_bdiv_qr_n
* documentation error wrt -ve exponents in mpz_powm_ui
* correct mpq_cmp_ui declaration error found by David Cleaver
* correct bug in Windows assembler code for karasub

Speedups:

* Completely new FFT implementation (William Hart)

Features:

* Capability to deal with full 32/64 bit words on 32/64 bit Windows (Brian Gladman)
* Python Windows build generator (Brian Gladman)

Compatibility:

* Add macros for __GNU_MP_RELEASE and __MPIR_RELEASE

Changes between MPIR 2.5.2 and MPIR 2.5.1

Bug Fixes:

* fixed build bug due to incorrect header in redc_2
* proper test for unpatched gcc-4.3.2

Speedups:

* Some new cpuids supported
* assembly redc_1 for sandybridge and bobcat for x86_64 and x86_64w

Compatibility:

* Added undocumented macros in gmp-impl.h for mpfr-3.1.0

Changes between MPIR 2.5.1 and MPIR 2.5.0

Bug Fixes:

 * Correct very rare errors in multiplication code
 * Correct very rare error in GCD code
 * minor MSVC build fixes

Features:

 * addadd addsub subadd sumdiff unconditionally availible on all systems 
 * redc_2 availible

Changes between MPIR 2.5.0 and MPIR 2.4.0

Bug Fixes:

 * Correct rounding mode make check tests

Speedups:

 * x86_64 fat build now uses RIP relative addressing
 * Toom22 speedup via assembler
 * New x86_64 assembler code
 * New Toom eval function using the new assembler code
 * Detect another Sandybridge CPU
 * New GCD code (by the GMP project)

Features:

 * General cleanup of old features 
 * Upgrade intenal component YASM
 * MSVC build tweeks
 * New experimental MSVC python build system

Changes:

 * Remove explicit support for thumb m68k and sh cpu's

Changes between MPIR 2.4.0 and MPIR 2.3.1

Bug Fixes:

 * The define __GMP_CC in mpir.h and gmp.h can miss out the the option
   -std=gnu99 
 * Old versions of GAS can fail to assemble the mpn/x86_64/core2/popcount.asm
 * Make install will fail to install the file gmpxx.h (only needed with the
   configure options --enable-cxx --enable-gmpcompat) when built out of the
   source tree.
 * Add an alias used by MPFR (snprintf -> _snprintf)
 * Correct incorrect parameter type for mpz_cdiv_q_2exp
 * Corrected CRLF endings in some C files
 * Corrected bitcnt_t functions error return on Win64 

Speedups:

 * Sandybridge and Bobcat have been tuned

memory.c
/* Memory allocation routines.

Copyright 1991, 1993, 1994, 2000, 2001, 2002 Free Software Foundation, Inc.

This file is part of the GNU MP Library.

The GNU MP Library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at your
option) any later version.

The GNU MP Library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with the GNU MP Library; see the file COPYING.LIB.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
MA 02110-1301, USA. */

#include <stdio.h>
#include <stdlib.h> /* for malloc, realloc, free */

#include "mpir.h"
#include "gmp-impl.h"


void *	(*__gmp_allocate_func)(size_t) = __gmp_default_allocate;
void *	(*__gmp_reallocate_func)(void *, size_t, size_t) = __gmp_default_reallocate;
void	(*__gmp_free_func)(void *, size_t) = __gmp_default_free;


/* Default allocation functions.  In case of failure to allocate/reallocate
   an error message is written to stderr and the program aborts.  */

void *
__gmp_default_allocate (size_t size)
{
  void *ret;
#ifdef DEBUG
  size_t req_size = size;
  size += 2 * BYTES_PER_MP_LIMB;
#endif
  ret = malloc (size);
  if (ret == 0)
    {
#if defined _MSC_VER && defined _WIN64
      fprintf (stderr, "GNU MP: Cannot allocate memory (size=%llu)\n", size);
#else
      fprintf (stderr, "GNU MP: Cannot allocate memory (size=%lu)\n", size);
#endif
      abort ();
    }

#ifdef DEBUG
  {
    mp_ptr p = ret;
    p++;
    p[-1] = (0xdeadbeef << 31) + 0xdeafdeed;
    if (req_size % BYTES_PER_MP_LIMB == 0)
      p[req_size / BYTES_PER_MP_LIMB] = ~((0xdeadbeef << 31) + 0xdeafdeed);
    ret = p;
  }
#endif
  return ret;
}

void *
__gmp_default_reallocate (void *oldptr, size_t old_size, size_t new_size)
{
  void *ret;

#ifdef DEBUG
  size_t req_size = new_size;

  if (old_size != 0)
    {
      mp_ptr p = oldptr;
      if (p[-1] != (0xdeadbeef << 31) + 0xdeafdeed)
	{
	  fprintf (stderr, "gmp: (realloc) data clobbered before allocation block\n");
	  abort ();
	}
      if (old_size % BYTES_PER_MP_LIMB == 0)
	if (p[old_size / BYTES_PER_MP_LIMB] != ~((0xdeadbeef << 31) + 0xdeafdeed))
	  {
	    fprintf (stderr, "gmp: (realloc) data clobbered after allocation block\n");
	    abort ();
	  }
      oldptr = p - 1;
    }

  new_size += 2 * BYTES_PER_MP_LIMB;
#endif

  ret = realloc (oldptr, new_size);
  if (ret == 0)
    {
#if defined _MSC_VER && defined _WIN64
        fprintf (stderr, "GNU MP: Cannot reallocate memory (old_size=%llu new_size=%llu)\n", old_size, new_size);
#else
        fprintf (stderr, "GNU MP: Cannot reallocate memory (old_size=%lu new_size=%lu)\n", old_size, new_size);
#endif
      abort ();
    }

#ifdef DEBUG
  {
    mp_ptr p = ret;
    p++;
    p[-1] = (0xdeadbeef << 31) + 0xdeafdeed;
    if (req_size % BYTES_PER_MP_LIMB == 0)
      p[req_size / BYTES_PER_MP_LIMB] = ~((0xdeadbeef << 31) + 0xdeafdeed);
    ret = p;
  }
#endif
  return ret;
}

void
__gmp_default_free (void *blk_ptr, size_t blk_size)
{
#ifdef DEBUG
  {
    mp_ptr p = blk_ptr;
    if (blk_size != 0)
      {
	if (p[-1] != (0xdeadbeef << 31) + 0xdeafdeed)
	  {
	    fprintf (stderr, "gmp: (free) data clobbered before allocation block\n");
	    abort ();
	  }
	if (blk_size % BYTES_PER_MP_LIMB == 0)
	  if (p[blk_size / BYTES_PER_MP_LIMB] != ~((0xdeadbeef << 31) + 0xdeafdeed))
	    {
	      fprintf (stderr, "gmp: (free) data clobbered after allocation block\n");
	      abort ();
	    }
      }
    blk_ptr = p - 1;
  }
#endif
  free (blk_ptr);
}
